### PR TITLE
[Validation] v8 blocks, Sapling merkle tree inclusion.

### PR DESCRIPTION
**Pull request by furszy**
Another decoupling from #1798, built on top of #1903 and #1915.

Containing three topics:

1) Sapling merkle tree root hash inclusion in the block header.
   * Block version bumped to v8.
   * Miner crafting the new merkle tree root hash and setting it into the block.

2) Track network total value entering and exiting the shielded pool. (*)
   * Similar to the previous, initial, zc failsafe flow. (this can be done much better, like we did with zerocoin supply)
   * Block negative shielded value pool balances. -- Network validation consensus rule --

3) ChainTip signal implemented.
   * Signal to notify the listeners (wallet) about new blocks (or chain reorgs), broadcasting the new block and the current merkle tree so the listener can build upon it. -- Not connected to anything yet --

#### This PR is including the following commits decoupled from #1798.

* Track net value entering and exiting the Sapling circuit.—> 8dd2045034b146d68eb6ef592433348852d54edd
* Block: header hashFinalSaplingRoot inclusion.—> 5f55a76bd303e21ed31e1947eb12d44a999f667f
* Sapling: missing nSaplingValue in LoadBlockIndexGuts added.—> fbe07613cf296ccc3f0a1b55f0015e86e2ff3d39
* Miner: sapling tree hash connection.—> cb88fae28eaa760c7feb00ab5ccd6f1cb8e6db15
* Implement ChainTip signal. —> 4e96e6c97d258c87910d2e9ab13cbb79c25ebda5
* Signal ChainTip in ConnectTip and DisconnectTip methods —> 85f4435b4076c8d4c8e222a0e0d7b8f76465be63
* Pass chainparams as argument in DisconnectTip, InvalidateBlock, ReprocessBlocks, DisconnectBlocks, DisconnectTip —> 1c809536b862e59554dc9ebee8651c56935a0f9a
* Do not try to update tree anchor and witnesses if sapling is not active --> 1c64497d4f020d6c400c7aa836f75dce7d1517d4
* Validation: Adapted the cached incremental witnesses update flow to the syncTransaction changes. —> 6e6372df9270dba2f76836d5b964c2153e28bda6

### CODE REVIEW

**Comment by furszy** [comment]
rebased this on top of #1915, making it depend on it.
The reason is that #1915 makes `ChainTip` signal flow much performant with cleaner code.

File: src/chain.h [hunk 1]
```diff
@@ -217,9 +217,19 @@ class CBlockIndex
...
     unsigned int nFlags{0};
 
+    //! Change in value held by the Sapling circuit over this block.
+    //! Not a boost::optional because this was added before Sapling activated, so we can
+    //! rely on the invariant that every block before this was added had nSaplingValue = 0.
+    CAmount nSaplingValue{0};
+
+    //! (memory only) Total value held by the Sapling circuit up to and including this block.
+    //! Will be boost::none if nChainTx is zero.
+    boost::optional<CAmount> nChainSaplingValue{boost::none};
```
Comment (Fuzzbawls) [inline src/chain.h:227]: If you include our `optional.h` header, can use our wrapper template and substitute
```cpp
Optional<CAmount> nChainSaplingValue{nullopt};
```

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -75,6 +75,25 @@ double GetDifficulty(const CBlockIndex* blockindex)
     return dDiff;
 }
 
+static UniValue ValuePoolDesc(
+        const std::string &name,
+        const boost::optional<CAmount> chainValue,
+        const boost::optional<CAmount> valueDelta)
```
Comment (Fuzzbawls) [inline src/rpc/blockchain.cpp:81]: Same here, include `optional.h` and use `Optional<CAmount> ...`

File: src/validation.cpp [hunk 3]
```diff
@@ -1913,6 +1931,12 @@ bool static DisconnectTip(CValidationState& state)
     for (const auto& tx : block.vtx) {
         GetMainSignals().SyncTransaction(*tx, pindexDelete->pprev, CMainSignals::SYNC_TRANSACTION_NOT_IN_BLOCK);
     }
+
+    if (chainparams.GetConsensus().NetworkUpgradeActive(pindexDelete->nHeight, Consensus::UPGRADE_V5_DUMMY)) {
+        // Update Sapling cached incremental witnesses
+        GetMainSignals().ChainTip(pindexDelete, &block, boost::none);
```
Comment (Fuzzbawls) [inline src/validation.cpp:1937]: `nullopt` instead of `boost::none`

File: src/validation.cpp [hunk 4]
```diff
@@ -2431,6 +2487,18 @@ bool ReceivedBlockTransactions(const CBlock& block, CValidationState& state, CBl
             CBlockIndex* pindex = queue.front();
             queue.pop_front();
             pindex->nChainTx = (pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx;
+
+            // Sapling, update chain value
+            if (pindex->pprev) {
+                if (pindex->pprev->nChainSaplingValue) {
+                    pindex->nChainSaplingValue = *pindex->pprev->nChainSaplingValue + pindex->nSaplingValue;
+                } else {
+                    pindex->nChainSaplingValue = boost::none;
```
Comment (Fuzzbawls) [inline src/validation.cpp:2496]: `nullopt` instead of `boost::none`

File: src/validation.cpp [hunk 5]
```diff
@@ -3490,12 +3558,21 @@ bool static LoadBlockIndexDB(std::string& strError)
             if (pindex->pprev) {
                 if (pindex->pprev->nChainTx) {
                     pindex->nChainTx = pindex->pprev->nChainTx + pindex->nTx;
+                    // Sapling, calculate chain index value
+                    if (pindex->pprev->nChainSaplingValue) {
+                        pindex->nChainSaplingValue = *pindex->pprev->nChainSaplingValue + pindex->nSaplingValue;
+                    } else {
+                        pindex->nChainSaplingValue = boost::none;
```
Comment (Fuzzbawls) [inline src/validation.cpp:3565]: `nullopt` instead of `boost::none`

File: src/validation.cpp [hunk 6]
```diff
@@ -3490,12 +3558,21 @@ bool static LoadBlockIndexDB(std::string& strError)
...
+                    }
+
                 } else {
                     pindex->nChainTx = 0;
+                    pindex->nChainSaplingValue = boost::none;
```
Comment (Fuzzbawls) [inline src/validation.cpp:3570]: `nullopt` instead of `boost::none`

File: src/validationinterface.h [hunk 7]
```diff
@@ -10,6 +10,8 @@
 #include <boost/signals2/signal.hpp>
 #include <boost/shared_ptr.hpp>
 
+#include "sapling/incrementalmerkletree.hpp"
```
Comment (Fuzzbawls) [inline src/validationinterface.h:13]: should be above the 3rd party lib includes

File: src/validationinterface.h [hunk 8]
```diff
@@ -33,6 +35,7 @@ class CValidationInterface {
 protected:
     virtual void UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {}
     virtual void SyncTransaction(const CTransaction &tx, const CBlockIndex *pindex, int posInBlock) {}
+    virtual void ChainTip(const CBlockIndex *pindex, const CBlock *pblock, boost::optional<SaplingMerkleTree> added) {}
```
Comment (Fuzzbawls) [inline src/validationinterface.h:38]: include `optional.h` and use `Optional<>` instead of `boost::optional<>`

File: src/validationinterface.h [hunk 9]
```diff
@@ -64,6 +67,9 @@ struct CMainSignals {
     boost::signals2::signal<void (const CBlock&, const CValidationState&)> BlockChecked;
     /** Notifies listeners that a block has been successfully mined */
     boost::signals2::signal<void (const uint256 &)> BlockFound;
+
+    /** Notifies listeners of a change to the tip of the active block chain. */
+    boost::signals2::signal<void (const CBlockIndex *, const CBlock *, boost::optional<SaplingMerkleTree>)> ChainTip;
```
Comment (Fuzzbawls) [inline src/validationinterface.h:72]: include `optional.h` and use `Optional<>` instead of `boost::optional<>`

**Review by Fuzzbawls — COMMENTED** [review]
Some consistency comments left about using our `optional.h` wrapper around `boost::optional`, but otherwise looking good

**Review by random-zebra — CHANGES_REQUESTED** [review]
There's an issue with the new RPC fields implementation (UniValue object elements must be a pair key-object, not plain objects).

Aside from this, we should consider if we really want to store in memory `nChainSaplingValue` for all blocks. That would add more than 20 MB to the runtime memory consumption at the current chain size (on mainnet). 
Not so much... but still, we only need its value at the chain tip.
Probably fine to leave it as is in this PR, but  something we might want to consider later.

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -75,6 +75,25 @@ double GetDifficulty(const CBlockIndex* blockindex)
...
+        const std::string &name,
+        const Optional<CAmount> chainValue,
+        const Optional<CAmount> valueDelta)
+{
+    UniValue rv(UniValue::VOBJ);
+    rv.pushKV("id", name);
+    rv.pushKV("monitored", (bool)chainValue);
+    if (chainValue) {
+        rv.pushKV("chainValue", ValueFromAmount(*chainValue));
+        rv.pushKV("chainValueZat", *chainValue);
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/blockchain.cpp:88]: yeah, done.
Plus, removed the `monitored` boolean as well.

File: src/rpc/blockchain.cpp [hunk 11]
```diff
@@ -137,6 +137,7 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* blockindex, bool tx
     result.pushKV("version", block.nVersion);
     result.pushKV("merkleroot", block.hashMerkleRoot.GetHex());
     result.pushKV("acc_checkpoint", block.nAccumulatorCheckpoint.GetHex());
+    result.pushKV("finalsaplingroot", block.hashFinalSaplingRoot.GetHex());
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/blockchain.cpp:140]: doing it. 
Curious thing, we never documented the accumulator checkpoint in `getblock`.

[the remark this answers is not in the document]
Comment (furszy) [reply src/rpc/blockchain.cpp:140]: And for the release notes, better to leave them for when all of the Sapling work gets merged.

All of the `Sapling` tagged PRs will need them. Up until now, haven't added anything to the release notes and we have merged several new features, RPC commands and consensus changes in master. So.. better to continue moving and do it all at once properly at the end.

File: src/validation.cpp [hunk 12]
```diff
@@ -1941,6 +1943,7 @@ struct ConnectTrace {
  */
 bool static ConnectTip(CValidationState& state, CBlockIndex* pindexNew, const std::shared_ptr<const CBlock>& pblock, bool fAlreadyChecked, ConnectTrace& connectTrace)
 {
+    const CChainParams& chainparams = Params();
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/validation.cpp:1968]: yeah, can squash them now. 
At the beginning the flow was totally different, #1915 made this quite nicer.

**Comment by furszy** [comment]
Updated per feedback.

File: src/rpc/blockchain.cpp [hunk 13]
```diff
@@ -563,6 +574,12 @@ UniValue getblockheader(const JSONRPCRequest& request)
             "  \"mediantime\" : ttt,    (numeric) The median block time in seconds since epoch (Jan 1 1970 GMT)\n"
             "  \"nonce\" : n,           (numeric) The nonce\n"
             "  \"bits\" : \"1d00ffff\", (string) The bits\n"
+            "  \"shielded_pool_value\": (object) Block shielded pool value\n"
+            "  {\n"
+            "     \"chainValue\":        (bool) Total value held by the Sapling circuit up to and including this block\n"
+            "     \"valueDelta\":        (bool) Change in value held by the Sapling circuit over this block\n"
```
Comment (random-zebra) [inline src/rpc/blockchain.cpp:582]: (e971fccc144386e16be0cc7ebf2ecbfcab0dc5e8)
numeric, not bool. Value expressed in PIV.

File: src/rpc/blockchain.cpp [hunk 14]
```diff
@@ -936,6 +953,11 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
             "  \"difficulty\": xxxxxx,     (numeric) the current difficulty\n"
             "  \"verificationprogress\": xxxx, (numeric) estimate of verification progress [0..1]\n"
             "  \"chainwork\": \"xxxx\"     (string) total amount of work in active chain, in hexadecimal\n"
+            "  \"shielded_pool_value\": (object) Chain tip shielded pool value\n"
+            "  {\n"
+            "     \"chainValue\":        (bool) Total value held by the Sapling circuit up to and including the chain tip\n"
+            "     \"valueDelta\":        (bool) Change in value held by the Sapling circuit over the chain tip block\n"
```
Comment (random-zebra) [inline src/rpc/blockchain.cpp:961]: (e971fccc144386e16be0cc7ebf2ecbfcab0dc5e8)
same

**Review by random-zebra — APPROVED** [review]
Just a minor nit for the help output.
Otherwise ACK.

**Comment by furszy** [comment]
Done, updated per feedback.

**Review by random-zebra — APPROVED** [review]
utACK 140391871c0a0b08d1e38cde464470ec71b6dddd

**Review by Fuzzbawls — APPROVED** [review]
utACK 140391871c0a0b08d1e38cde464470ec71b6dddd

**Pull request merged**